### PR TITLE
dev/core#890 - Multiple line item shown on view contribution if parti…

### PR DESCRIPTION
**Pull request by jitendrapurohit**
…cipant is transferred to another contact

Overview
----------------------------------------
Reopened https://github.com/civicrm/civicrm-core/pull/14104 with some changes to financial item entries as per https://github.com/civicrm/civicrm-core/pull/14104#issuecomment-586675243 (Solution 1 proposal).

Multiple line item shown on view contribution if participant is transferred to another contact

Before
----------------------------------------
To replicate -

- Create a priceset for an event with `Text / Numeric Quantity` field type. Amount = $10.
- Register contact A to the event using the above price option.
- Transfer the event participation to another contact B.
- View contribution on the main contact, it displays double line items for the participation. 

image

I think the older participant status is set to "transferred" so it should remove the contribution link from the line item table?

After
----------------------------------------

Line item is displayed correctly.

image

Comments
----------------------------------------
Gitlab - https://lab.civicrm.org/dev/core/issues/890

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit the test fail relates

**Comment by jitendrapurohit** [comment]
Have updated the PR with the test fix @eileenmcnaughton

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit I tested this with a different field type and got an error :-(

Screen Shot 2020-04-04 at 2 25 57 PM

Since I worked through the code change I pulled out the extraction into a separate PR to get that merged & make it easier to work through this - https://github.com/civicrm/civicrm-core/pull/16976

**Comment by jitendrapurohit** [comment]
@eileenmcnaughton Have updated the PR and fixed the test. I tried to replicate the above db error but wasn't able to see any problems with different combinations of fields. 

image

From the backtrace, it seems event id was not retrieved on your side. This PR also modifies its value so it should already be handled. Can you pls confirm if you still notice the above error?

Thanks.

File: CRM/Event/Form/SelfSvcTransfer.php [hunk 1]
```diff
@@ -247,8 +247,6 @@ public static function checkProfileComplete($fields, &$errors, $self) {
     }
     if (!$email && !(CRM_Utils_Array::value('first_name', $fields) &&
       CRM_Utils_Array::value('last_name', $fields))) {
-      $defaults = $params = ['id' => $eventId];
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/SelfSvcTransfer.php:250]: Why is this removed?

File: CRM/Event/Form/SelfSvcTransfer.php [hunk 2]
```diff
@@ -500,17 +498,15 @@ public function sendCancellation() {
...
    * @param int $toContactID
-   * @param int $fromParticipantID
-   * @param int $eventID
    *
    * @throws \CRM_Core_Exception
    * @throws \CiviCRM_API3_Exception
    */
-  public function transferParticipantRegistration($toContactID, $fromParticipantID, $eventID) {
-    $query = 'select role_id, source, fee_level, is_test, is_pay_later, fee_amount, discount_id, fee_currency,campaign_id, discount_amount from civicrm_participant where id = ' . $fromParticipantID;
+  protected function transferParticipantRegistration($toContactID) {
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/SelfSvcTransfer.php:508]: So  in terms of where we are going with this code - I'm hoping it would be moved to the BAO at some point as a move() function & called from the api. I mention this because this change makes the function more rather than less tied to the form it is on - which is the opposite direction

File: CRM/Event/Form/SelfSvcTransfer.php [hunk 3]
```diff
@@ -522,7 +518,7 @@ public function transferParticipantRegistration($toContactID, $fromParticipantID
       $value_to['fee_amount'] = $dao->fee_amount;
     }
     $value_to['contact_id'] = $toContactID;
-    $value_to['event_id'] = $eventID;
+    $value_to['event_id'] = $this->_event_id;
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/SelfSvcTransfer.php:521]: given the above query we should probably just get event_id in the query

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit OK - the fatal errors occur when there is no associated contribution

File: CRM/Event/Form/SelfSvcTransfer.php (hunk 1 above)
Comment (jitendrapurohit) [reply CRM/Event/Form/SelfSvcTransfer.php:250]: Unnecessary lines. `$eventId` is an undeclared variable.

File: CRM/Event/Form/SelfSvcTransfer.php (hunk 2 above)
Comment (jitendrapurohit) [reply CRM/Event/Form/SelfSvcTransfer.php]: cool. I've updated the PR to keep the fromParticipantID param :)

File: CRM/Event/Form/SelfSvcTransfer.php (hunk 3 above)
Comment (jitendrapurohit) [reply CRM/Event/Form/SelfSvcTransfer.php]: done. Have updated with the api4 call and created the toParticipant with the retrieved params 👍

**Comment by jaapjansma** [comment]
test this please

**Comment by jaapjansma** [comment]
@jitendrapurohit can you have a look at the failing test?

**Comment by jitendrapurohit** [comment]
Hmm, it looks like the [details](https://test.civicrm.org/job/CiviCRM-Core-PR/33503/) page does not display any specific test which is failing.

From the console output https://test.civicrm.org/job/CiviCRM-Core-PR/33503/console, It seems the failure is resulted during the execution of EventTest::testTemplateFilterByDefault function?

>not ok 1 - Error: EventTest::testTemplateFilterByDefault

>API_Exception: Unknown api parameter: getHaving in /home/jenkins/bknix-dfl/build/core-16956-mxxj/web/sites/all/modules/civicrm/Civi/Api4/Generic/AbstractAction.php on line 215
  
I tried executing the same on my local and it displays a success message on completion.

    jitendra$ phpunit5 tests/phpunit/api/v4/Action/EventTest.php           
    Parsing schema description /Users/jitendra/src/civicrm/xml/schema/Schema.xml
    Extracting database information
    Extracting table information
    PHPUnit 5.7.27 by Sebastian Bergmann and contributors.
    
    .                                                                   1 / 1 (100%)
    
    Time: 1.18 seconds, Memory: 28.00MB
    
    OK (1 test, 1 assertion)

**Comment by jitendrapurohit** [comment]
retest this please

**Comment by jitendrapurohit** [comment]
test this please

**Comment by jaapjansma** [comment]
@jitendrapurohit are you able to fix the conflicting files?

**Comment by jitendrapurohit** [comment]
@jaapjansma Sorry for the delay. I've now updated the PR with the conflict fixes. Thanks.

**Comment by jitendrapurohit** [comment]
@jaapjansma can you pls remove the `needs work` label from this PR? This was updated with the conflict fixes and also passes the jenkin test process.

**Comment by eileenmcnaughton** [comment]
I re-tested this & it seems to work fine. Issues from previous comments have been addressed. Merging
